pytorch_lightning/loops/batch/training_batch_loop.py

```diff
@@ -98,13 +98,10 @@ def on_run_end(self) -> _OUTPUTS_TYPE:
         self.optimizer_loop._hiddens = None
         # this is not necessary as the manual loop runs for only 1 iteration, but just in case
         self.manual_loop._hiddens = None
-        output, self._outputs = self._outputs, None  # free memory
+        output, self._outputs = self._outputs, []  # free memory
+        self._remaining_splits = []
         return output
 
-    def teardown(self) -> None:
-        # release memory
-        self._remaining_splits = None
```

Comment:
we still need to free memory when on_run_end didn't run. That's why I put it here! 
what we forgot here is to call `optimizer_loop.teardown()`

Reply:
`_remaining_splits` is filled in `on_run_start`, which means we are free to clean it in `on_run_end`. why wait until `teardown` to do it?